### ISSUE TITLE
Avoid C files with the same name to simplify libr.a

### ISSUE THREAD

**Issue by radare** [radare/radare2#11878]
I cooked a script to make a list of all the files with duplicated filenames. this will basically allow us to un-ar all the archives in a single directory and get a single libr.a that works without all the non-portable/messy/hacky partial-ld thing that is breaking static builds on many platforms.

Another benefit of doing this is that static build errors will be easier to understand because a single .o file will refer to a unique object file. Having multiple objects with the same name in library archives is "useful" but the root of all evil

```
$ node cfiles.js
DUPE dis.c
DUPE asm.c
DUPE disasm.c
DUPE log.c
DUPE spaces.c
DUPE format.c
DUPE graph.c
DUPE wasm.c
DUPE main.c
DUPE print.c
DUPE calc.c
DUPE code.c
DUPE plugin.c
DUPE cache.c
DUPE lib.c
DUPE file.c
DUPE pipe.c
DUPE utf8.c
DUPE diff.c
DUPE state.c
DUPE cond.c
DUPE op.c
DUPE xrefs.c
DUPE value.c
DUPE session.c
DUPE esil.c
DUPE map.c
DUPE signal.c
DUPE darwin.c
DUPE reg.c
DUPE core.c
DUPE utils.c
DUPE packet.c
DUPE adler32.c
DUPE test.c
DUPE list.c
DUPE fs.c
DUPE disk.c
```

The script:
```js
const fs = require('fs');

const cfiles = fs.readFileSync('cfiles').toString();

const cnames = cfiles.split('\n').map(_ => {
  var a = _.split('/');
  return a[a.length - 1];
});

const cache = {};
const duped = {};
for (let cfile of cnames) {
	if (cache[cfile]) {
		if (!duped[cfile]) {
			console.error("DUPE", cfile);
		}
		duped[cfile] = true;
	}
	cache[cfile] = true;
}
```

How to run
```sh
find libr shlr| grep '\.c$' > cfiles
node cfiles.js
```

**Comment by radare** [comment]
The way im buliding the static libs right now works fine because im packing each module in a separate directory, so as long as there are no duplicated filenames on any module it will work.

**Comment by radare** [comment]
No issues on this anymore, the current build system for creating static libraries is working fine and its been tested for a while. we may be able to detect when this happens again thanks to the radare2-release repo, so imho we can close it.

**Comment by XVilka** [comment]
@ret2libc you might be interested in this as well.

**Comment by ret2libc** [comment]
> @ret2libc you might be interested in this as well.

Why?

I think meson doesn't really use `libr.a` and I (still) don't understand the purpose of it.

**Comment by trufae** [comment]
i fixed this long time ago (no two .c files with the same name), actually when i wrote the script. Let me rephrase:

* If filenames are unique across the project it is easier for everyone if `reg.c` relates to libr_reg, libr_debug, libr_anal or libr_core.
* Archive files support multiple files with the same name, but this is problematic because the way to handle that depends on the version of AR and the way its used. this is why right now it is built this way
* the idea for this issue is to have a way to detect this at PR time
* Meson does not generate libr.a which is imho much better way to use r2 on 3rd party projects (r2-sdk with libr.a and libr.dylib/so/dll) instead of having to drop 50 files to get r2 working
